app/models/concerns/smart_proxy_host_extensions.rb

```diff
@@ -0,0 +1,52 @@
+module SmartProxyHostExtensions
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def smart_proxy_reference(hash)
+      SmartProxyReference.add_smart_proxy_reference hash
+    end
+
+    def smart_proxy_ids(hosts_scope = Host::Managed.where(nil))
+      hosts_scope.eager_load(proxy_join_tables).pluck(proxy_column_list).flatten.uniq.compact
+    end
+
+    def proxy_self_columns
+      proxy_join_references(true).values.first.map(&:to_s)
+    end
+
+    def proxy_join_tables
+      proxy_join_references.keys
+    end
+
+    def create_groups
```

Comment:
Pretty late in the review party, but since the module is getting included in to `Host` class, this name can be pretty misleading in the context of the host. Could you think of a better name?